src/ssl/gadgets.cc

```diff
@@ -484,8 +505,16 @@ addAltNameWithSubjectCn(Security::CertPointer &cert)
     if (!cn_data)
         return false;
 
+    sockaddr_storage socket_info;
+    const char* altname_type;
+    const int is_ip = Ssl::is_ip_address(cn_data->data, &socket_info);
```

Comment:
Please see if you can use Ip::Address::operator "=" instead of calling is_ip_address() here. Ip::Address::operator "=" API is terrible, but it is existing code that we do not want to semi-duplicate unless absolutely necessary (and we would have to semi-duplicate it differently if we have to do that at all).

N.B. This PR adds is_ip_address() that uses inet_pton(). Latest Squid code does not use inet_pton(). It uses Ip::Address::lookupHostIP(...,true) callers. That private Ip::Address method currently uses getaddrinfo().

Reply:
Hi @rousskov. I wonder if you could help please. I'm attempting to use `Ip::Address::operator` by adding `#include ip/Address.h` to gadgets.cc. I then do something like this:
```
Ip::Address ipAddress;
if (ipAddress = address_str) {
  type = "IP"
} else {
  type = "DNS"
}
```
After running `make` I get this error in the output:

```
/usr/bin/ld: ../../../../src/ssl/.libs/libsslutil.a(gadgets.o): in function `Ip::Address::Address()':
/tmp/squid-6.9/src/ssl/../../src/ip/Address.h:46: undefined reference to `Ip::Address::setEmpty()'
/usr/bin/ld: ../../../../src/ssl/.libs/libsslutil.a(gadgets.o): in function `addAltNameWithSubjectCn(Security::LockingPointer<x509_st, &Security::X509_free_cpp, HardFun<int, x509_st*, &X509_up_ref> >&)':
/tmp/squid-6.9/src/ssl/gadgets.cc:490: undefined reference to `Ip::Address::operator=(char const*)'
collect2: error: ld returned 1 exit status
```

I've tried to use a couple of the `Ip::Address` methods which resulted in the same error. Would you be able to point me in the right direction to fix this please?

Thanks

Reply:
Shared output does not detail which executable fails to link, but I am guessing it is security_file_certgen. See the difference between two comments describing two libraries in src/ssl/Makefile.am. Compared to the primary squid executable, security_file_certgen helper is build using a (smaller but naturally growing) subset of libraries; we split src/ssl code into two libraries to minimize the number of (indirect) security_file_certgen dependencies[^1].

If my guess is correct, then try adding `$(top_builddir)/src/ip/libip.la` to `security_file_certgen_LDADD` in src/security/cert_generators/file/Makefile.am. It should probably go on the first or second line, above or below libsslutil.la. You may they need to add more libraries if libip.la has unresolved dependencies. Do not forget to re-bootstrap after modifying an Automake file!

If the suggestion above works, then the original two-library split described above will no longer be necessary (but removing it should be done in a dedicated PR so do not worry about it for now).

[^1]: We plan to migrate away from that painful minimization approach, but the code you have to deal with has not been upgraded, so you have to link security_file_certgen (and other binaries) with the necessary convenience libraries (often discovered by looking at linker errors like the one you have shared).

Reply:
You were right. However, I've made the change you suggested to `src/security/cert_generators/file/Makefile.am` then I ran `./bootstrap.sh` followed by `./configure` from `$(top_builddir)` but I see no change in the error reported. Here it is with a bit more context:

```
make[4]: Entering directory '/portal/squid-6.9/src/security/cert_generators'                                                      
Making all in file                                                                                                                
make[5]: Entering directory '/portal/squid-6.9/src/security/cert_generators/file'
/bin/bash ../../../../libtool  --tag=CXX   --mode=link g++ -std=c++17 -Wall -Wextra -Wimplicit-fallthrough=5 -Wpointer-arith -Wwrite-strings -Wcomments -Wshadow -Wmissing-declarations -Woverloaded-virtual -Werror -pipe -D_REENTRANT -I/usr/include/p11-kit-1    -g -O2 -march=native  -g -o security_file_certgen certificate_db.o security_file_certgen.o ../../../../src/ip/libip.la ../../../../src/ssl/libsslutil.la ../../../../src/sbuf/libsbuf.la ../../../../src/debug/libdebug.la ../../../../src/error/liberror.la ../../../../src/comm/libminimal.la ../../../../src/mem/libminimal.la ../../../../src/base/libbase.la ../../../../src/time/libtime.la -lssl -lcrypto   -lgnutls   ../../../../compat/libcompatsquid.la                                                                      
libtool: link: g++ -std=c++17 -Wall -Wextra -Wimplicit-fallthrough=5 -Wpointer-arith -Wwrite-strings -Wcomments -Wshadow -Wmissing-declarations -Woverloaded-virtual -Werror -pipe -D_REENTRANT -I/usr/include/p11-kit-1 -g -O2 -march=native -g -o security_file_certgen certificate_db.o security_file_certgen.o  ../../../../src/ip/.libs/libip.a ../../../../src/ssl/.libs/libsslutil.a ../../../../src/sbuf/.libs/libsbuf.a ../../../../src/debug/.libs/libdebug.a ../../../../src/error/.libs/liberror.a ../../../../src/comm/.libs/libminimal.a ../../../../src/mem/.libs/libminimal.a ../../../../src/base/.libs/libbase.a ../../../../src/time/.libs/libtime.a -lssl -lcrypto -lgnutls ../../../../compat/.libs/libcompatsquid.a                                                                   
/usr/bin/ld: ../../../../src/ssl/.libs/libsslutil.a(gadgets.o): in function `Ip::Address::Address()':
/portal/squid-6.9/src/ssl/../../src/ip/Address.h:46: undefined reference to `Ip::Address::setEmpty()'
/usr/bin/ld: ../../../../src/ssl/.libs/libsslutil.a(gadgets.o): in function `addAltNameWithSubjectCn(Security::LockingPointer<x509_st, &Security::X509_free_cpp, HardFun<int, x509_st*, &X509_up_ref> >&)':
/portal/squid-6.9/src/ssl/gadgets.cc:490: undefined reference to `Ip::Address::operator=(char const*)'
collect2: error: ld returned 1 exit status        
```

It seems like libsslutil is not seeing the link but I'm afraid I'm not sure how to proceed from here either if you can suggest please. Appreciate your help/patience.

Reply:
```
... security_file_certgen.o ../../../../src/ip/libip.la ../../../../src/ssl/libsslutil.la ...
```

Try reordering the first two libraries. In other words, place libip _after_ libsslutil in Makefile.am.

Rule of thumb: Libraries that provide symbol X should be listed _after_ libraries that use symbol X.

Reply:
Ah thanks. I had tried it below with a different source checkout and was sure I'd seen errors but doing as you suggest I was able to get `make` to complete. I'll now see about making some of these changes. Thanks.